### DISCUSSION TITLE
Static within Dynamic Branching

### DISCUSSION THREAD

**Discussion by stuvet** [ropensci/targets#544] [General]
I have been enjoying using targets for a few months now. Many thanks for an excellent package!

I may be way off base here, but I'm wondering how I might implement static within dynamic branching (I can do the reverse).

Use-case - cmdstan processing pipeline with threaded chains based on multiple datasets & I want to fit a different model for each dataset, repeating monthly. I'm using transient Slurm workers which can be preempted, & preemption causes failure of the pipeline, even with error = 'continue'.

To manage the threaded chains I'm running each chain on a separate worker then combining them by pattern = map(model, dataset). So far I can use tar_map & tar_combine to statically map each chain & combine the results by dataset, but if one machine is preempted the whole of the incomplete static branch is brought down & I lose the work done. This can be quite significant. Same problem for dynamic branching across datasets & chains & combining with a tar_group_by.

Ideally I'd statically branch across the dynamically-generated dataset ids using tar_map (and I'd probably set up a nested tar_map for the chains). That way I _think_ I'd only lose the work done on a single dataset if one of the chains (machines) is preempted.

Am I way off base here? Thoughts on how this might be accomplished?

Thanks,

Stuart

**Comment by wlandau** [comment]
I think there are simpler solutions to the problem. Have you tried [`stantargets`](https://docs.ropensci.org/stantargets/)? [`tar_stan_mcmc()`](https://docs.ropensci.org/stantargets/reference/tar_stan_mcmc.html) applies multiple models to a single dataset, and [`tarchetypes::tar_eval()`](https://books.ropensci.org/targets/static.html#metaprogramming) can generate repeated calls to `tar_stan_mcmc()` (one per dataset). The `parallel_chains` argument of `tar_stan_mcmc()` parallelizes chains across processes, `threads_per_chain` controls multithreading, and you can [configure the template file](https://books.ropensci.org/targets/hpc.html#clustermq-configuration) (either `clustemrq` or `future`/`bachtools`) to control the number of cores per SLURM worker.

Static branching within dynamic branching is impossible: static branching happens when the pipeline is *defined*, whereas dynamic branching happens when the pipeline is *executed*.

**Comment by stuvet** [comment]
Thanks for the fast response, & it’s very helpful to know it’s impossible
before I spend too much time on it!

Yes I tried stantargets for a bit, but I had some suspected
hardware-specific issues getting it to run models. I’m used to using brms
with a cmdstanr backend & I couldn’t get stantargets to run the models even
if produced by brms, & even if precompiled. Likely the solution is as
simple as fixing some stanc flags as using brms to call cmdstanr works
fine. I tried it when new to targets & struggled to understand how to
generalise the examples to my use case rather than fitting models to
multiple simulated datasets. I’ll probably have a better idea now,
especially with your pointers.

I’ll take another look & work on including this instead. Thanks for your
help.

Regards,

Stuart

**Reply by wlandau** [reply]
Some aspects of `brms` are actually a bit difficult to work with in `stantargets`, re https://github.com/ropensci/stantargets/issues/12. If you don't use `stantargets`, it still may be helpful to look at https://docs.ropensci.org/stantargets/articles/introduction.html#multiple-models-1 to understand that approach to pipeline construction. The compilation issues do sound solvable though because `tar_stan_mcmc()` accepts all the arguments of `cmdstanr::cmstan_model()`, including compiler flags.

**Comment by stuvet** [comment]
That’s incredibly helpful thanks Will. I hadn’t found these. I’ll take a
good look tomorrow & see if I can adapt my pipeline to suit.

**Reply by stuvet** [reply]
Here's a working example that could fit my purposes  - specifically to fit a series of brms-generated models to a dynamic series of datasets.

It adapts code from [ropensci/stantargets#12](https://github.com/ropensci/stantargets/issues/12)

I still need to check how reliably it detects outdated data from the datasets.RDS in real-world use, but my plan is to alter the pre-targets code to include a function which keeps this file up-to-date, likely to include file timestamps also.

I'm adding it here for anyone else who will try to do this - hope it helps someone.

UPDATE:
Added recovery of brmsfit object. Sorry for any mistakes with copypasta


```
# _targets.R
library(targets)
library(stantargets)
library(fs)
library(dplyr)

# output_dir for tar_stan_mcmc - needed to recover to brmsfit
output_dir <- '_scratch'
if(!dir.exists(output_dir)) dir.create(output_dir)


# Helper function - https://github.com/ropensci/stantargets/issues/12#issuecomment-747992278
make_standata_target <- function(stan_model, ...){
  if(!("brmsmodel" %in% class(stan_model)))
    stop("Expecting an object of class brmsmodel in stan_model argument")

  standata <- brms::make_standata(...)
  return(unclass(standata))
}

# Helper function - recovers cmdstanr models to brmsfit
as_brmsfit <- function(name_str, id_str, brms_wrapper, stanfit_sym) {
    wrapper <- brms_wrapper$fit[ brms_wrapper$id == id_str & brms_wrapper$name == name_str ][[1]]
    cmdstanfit <- stanfit_sym
    wrapper$fit <- rstan::read_stan_csv(cmdstanfit$output_files())
    brms:::rename_pars(wrapper)
}

# Generate initial dataset ids - could  be read from file or url.
# Done this way to test outdated target detection
dataset_meta_path<-'datasets.RDS'
if( !file.exists(dataset_meta_path)) saveRDS(c( "4f31061e", "6dbfe592" ), dataset_meta_path)
dataset_ids <- readRDS(dataset_meta_path)

# Randomly add an extra dataset
dataset_ids <- if( runif(1) > 0.3 ) {
                    print( 'No datasets added.' )
                    dataset_ids
                  } else {
                    print( 'Added one dataset!' )
                    c( dataset_ids, paste0( sample( c( 0:9, letters[1:6]), 8, replace = T), collapse = '' ) )
                  }

saveRDS(dataset_ids, 'datasets.RDS')

# Prepare stan_model files
model_names <- c('model1', 'model2')
stan_model_paths <- fs::file_create( paste0( model_names, '.stan' ))

# Prepare jobs grid for tar_eval
jobs <- expand_grid( id_str = dataset_ids, name_str = model_names) %>%
                  dplyr::mutate(
                    fit_str = sprintf( 'fit_%s_%s', name_str, id_str),
                    brmsfit_str = paste0( fit_str, '_brms'),
                    stanfit_sym = purrr::map( paste0( fit_str, '_mcmc_', name_str ), rlang::sym ),
                    path_str = paste0( name_str, '.stan' )
                  )

list(
   # Done to allow branching across dataset_ids. Probably a better way!!
   tar_target(
    dataset_meta,
    dataset_ids
  ),
  tar_target(
    datasets,
    # ... fetch list of datasets of unknown length e.g. fetch from cloud bucket.
    tibble(id = as.character(dataset_meta), data = list(tibble(z = rnorm(10), x = rnorm(10), y=  0.5 * rnorm(10) + 0.75 * rnorm(10)))),
    pattern = map(dataset_meta)
  ),
  tar_target(
    stan_spec,
    tibble(
      name = model_names,
      path = stan_model_paths,
      formula = c(
                            list(brmsformula(formula = y ~ 1 + x)), 
                            list(brmsformula(formula = y ~ 1 + x + z))
                        )
#    prior = # Required - full prior spec

    )
  ),
  # generates a single stan model per stan_spec.
  # Need to provide all priors etc to make a universal generated model, or adapt this to create a stan model for each dataset.
  tar_target(
    stan_model,
    make_stancode(
      formula = stan_spec$formula[[1]],
      data = datasets$data[[1]], # Used as demo data to generate single model for all datasets.
      save_model = stan_spec$path
#      prior = stan_spec$prior[[1]] ## Full prior required if aiming to share a single model across datasets
    ),
    pattern = map(stan_spec)
  ),
  tar_target(
    stan_data,
    tibble(
      id = datasets$id,
      name = stan_spec$name,
      data = list( 
                       make_standata_target(
                            stan_model,
                            formula = stan_spec$formula[[1]],
                            data = datasets$data
#                            prior = stan_spec$prior[[1]]  ## Required
                        )
                 )
    ),
    pattern = cross( datasets, map( stan_spec, stan_model ) )
  ),
  tar_target(
    brms_wrapper,
    tibble(
      id = datasets$id,
      name = stan_spec$name,
      fit  = list(
                brms::brm(
                    stan_spec$formula[[1]],
                    data = datasets$data,
#                  prior = stan_spec$prior[[1]], ## Required above
                    backend = 'cmdstanr',
                    empty = TRUE
                )
            )
    ),
    pattern = cross(datasets, map(stan_spec))
  ),
  tarchetypes::tar_eval(
      list(
        stantargets::tar_stan_mcmc(
            fit,
            stan_files = path_str,
            data = stan_data$data[ stan_data$id == id_str & stan_data$name == name_str ][[1]],
            refresh = 0,
            init = 1,
            output_dir = output_dir
        ),
        tar_target(
            brmsfit_str,
            {
                wrapper <- brms_wrapper$fit[ brms_wrapper$id == id_str & brms_wrapper$name == name_str ][[1]]
                model <- stanfit_sym
                wrapper$fit <- rstan::read_stan_csv( model$output_files() )
                brms:::rename_pars( wrapper )
           },
           packages = c('rstan', 'brms'),
        )
    ),
    values = jobs
  )
)
```

![image](https://user-images.githubusercontent.com/18164223/124962997-30a63280-e017-11eb-90b9-77c867f7460e.png)

Closeup of model branches to include brms

![image](https://user-images.githubusercontent.com/18164223/124963087-4b78a700-e017-11eb-9518-f2bdfeb14ba7.png)

**Reply by stuvet** [reply]
Updated to include recovery to brmsfit
